diff -x Makefile.org.orig -x Makefile.rej -x Makefile.orig -Nur openssl-1.0.0e.orig/Configure openssl-1.0.0e/Configure
--- openssl-1.0.0e.orig/Configure	2013-01-19 15:39:37 +0000
+++ openssl-1.0.0e/Configure	2013-01-19 15:49:07 +0000
@@ -1514,6 +1514,7 @@
 	s/^PLATFORM=.*$/PLATFORM=$target/;
 	s/^OPTIONS=.*$/OPTIONS=$options/;
 	s/^CONFIGURE_ARGS=.*$/CONFIGURE_ARGS=$argvstring/;
+	s/^ENGINESDIR=.*$/ENGINESDIR=$enginesdir/;
 	if ($cross_compile_prefix)
 		{
 		s/^CC=.*$/CROSS_COMPILE= $cross_compile_prefix\nCC= \$\(CROSS_COMPILE\)$cc/;
diff -x Makefile.org.orig -x Makefile.rej -x Makefile.orig -Nur openssl-1.0.0e.orig/Makefile openssl-1.0.0e/Makefile
--- openssl-1.0.0e.orig/Makefile	2013-01-19 15:39:40 +0000
+++ openssl-1.0.0e/Makefile	2013-01-19 15:49:07 +0000
@@ -15,6 +15,7 @@
 PLATFORM=dist
 OPTIONS= no-gmp no-jpake no-krb5 no-md2 no-rc5 no-rfc3779 no-shared no-store no-zlib no-zlib-dynamic static-engine
 CONFIGURE_ARGS=dist
+ENGINESDIR=
 SHLIB_TARGET=
 
 # HERE indicates where this Makefile lives.  This can be used to indicate
@@ -187,7 +188,7 @@
 		SDIRS='$(SDIRS)' LIBRPATH='$(INSTALLTOP)/$(LIBDIR)'	\
 		INSTALL_PREFIX='$(INSTALL_PREFIX)'		\
 		INSTALLTOP='$(INSTALLTOP)' OPENSSLDIR='$(OPENSSLDIR)'	\
-		LIBDIR='$(LIBDIR)'				\
+		LIBDIR='$(LIBDIR)' ENGINESDIR='$(ENGINESDIR)'	\
 		MAKEDEPEND='$$$${TOP}/util/domd $$$${TOP} -MD $(MAKEDEPPROG)' \
 		DEPFLAG='-DOPENSSL_NO_DEPRECATED $(DEPFLAG)'	\
 		MAKEDEPPROG='$(MAKEDEPPROG)'			\
@@ -499,7 +500,7 @@
 install_sw:
 	@$(PERL) $(TOP)/util/mkdir-p.pl $(INSTALL_PREFIX)$(INSTALLTOP)/bin \
 		$(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR) \
-		$(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines \
+		$(INSTALL_PREFIX)$(ENGINESDIR) \
 		$(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/pkgconfig \
 		$(INSTALL_PREFIX)$(INSTALLTOP)/include/openssl \
 		$(INSTALL_PREFIX)$(OPENSSLDIR)/misc \
diff -x Makefile.org.orig -x Makefile.rej -x Makefile.orig -Nur openssl-1.0.0e.orig/Makefile.org openssl-1.0.0e/Makefile.org
--- openssl-1.0.0e.orig/Makefile.org	2013-01-19 15:39:40 +0000
+++ openssl-1.0.0e/Makefile.org	2013-01-19 15:49:07 +0000
@@ -13,6 +13,7 @@
 PLATFORM=dist
 OPTIONS=
 CONFIGURE_ARGS=
+ENGINESDIR=
 SHLIB_TARGET=
 
 # HERE indicates where this Makefile lives.  This can be used to indicate
@@ -185,7 +186,7 @@
 		SDIRS='$(SDIRS)' LIBRPATH='$(INSTALLTOP)/$(LIBDIR)'	\
 		INSTALL_PREFIX='$(INSTALL_PREFIX)'		\
 		INSTALLTOP='$(INSTALLTOP)' OPENSSLDIR='$(OPENSSLDIR)'	\
-		LIBDIR='$(LIBDIR)'				\
+		LIBDIR='$(LIBDIR)' ENGINESDIR='$(ENGINESDIR)'	\
 		MAKEDEPEND='$$$${TOP}/util/domd $$$${TOP} -MD $(MAKEDEPPROG)' \
 		DEPFLAG='-DOPENSSL_NO_DEPRECATED $(DEPFLAG)'	\
 		MAKEDEPPROG='$(MAKEDEPPROG)'			\
@@ -500,7 +501,7 @@
 install_sw:
 	@$(PERL) $(TOP)/util/mkdir-p.pl $(INSTALL_PREFIX)$(INSTALLTOP)/bin \
 		$(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR) \
-		$(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines \
+		$(INSTALL_PREFIX)$(ENGINESDIR) \
 		$(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/pkgconfig \
 		$(INSTALL_PREFIX)$(INSTALLTOP)/include/openssl \
 		$(INSTALL_PREFIX)$(OPENSSLDIR)/misc \
diff -x Makefile.org.orig -x Makefile.rej -x Makefile.orig -Nur openssl-1.0.0e.orig/engines/Makefile openssl-1.0.0e/engines/Makefile
--- openssl-1.0.0e.orig/engines/Makefile	2013-01-19 15:39:39 +0000
+++ openssl-1.0.0e/engines/Makefile	2013-01-19 15:53:29 +0000
@@ -107,7 +107,7 @@
 	@[ -n "$(INSTALLTOP)" ] # should be set by top Makefile...
 	@if [ -n "$(SHARED_LIBS)" ]; then \
 		set -e; \
-		$(PERL) $(TOP)/util/mkdir-p.pl $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines; \
+		$(PERL) $(TOP)/util/mkdir-p.pl $(INSTALL_PREFIX)$(ENGINESDIR); \
 		for l in $(LIBNAMES); do \
 			( echo installing $$l; \
 			  pfx=lib; \
@@ -125,14 +125,14 @@
 				*DSO_BEOS*)	sfx=".so";;	\
 				*DSO_DLFCN*)	sfx=`expr "$(SHLIB_EXT)" : '.*\(\.[a-z][a-z]*\)' \| ".so"`;;	\
 				*DSO_DL*)	sfx=".sl";;	\
-				*DSO_WIN32*)	sfx="eay32.dll"; pfx=;;	\
+				*DSO_WIN32*)	sfx=".dll";;	\
 				*)		sfx=".bad";;	\
 				esac; \
-				cp $$pfx$$l$$sfx $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$$pfx$$l$$sfx.new; \
+				cp $$pfx$$l$$sfx $(INSTALL_PREFIX)$(ENGINESDIR)/$$pfx$$l$$sfx.new; \
 				;; \
 			  esac; \
-			  chmod 555 $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$$pfx$$l$$sfx.new; \
-			  mv -f $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$$pfx$$l$$sfx.new $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$$pfx$$l$$sfx ); \
+			  chmod 555 $(INSTALL_PREFIX)$(ENGINESDIR)/$$pfx$$l$$sfx.new; \
+			  mv -f $(INSTALL_PREFIX)$(ENGINESDIR)/$$pfx$$l$$sfx.new $(INSTALL_PREFIX)$(ENGINESDIR)/$$pfx$$l$$sfx ); \
 		done; \
 	fi
 	@target=install; $(RECURSIVE_MAKE)
diff -x Makefile.org.orig -x Makefile.rej -x Makefile.orig -Nur openssl-1.0.0e.orig/engines/ccgost/Makefile openssl-1.0.0e/engines/ccgost/Makefile
--- openssl-1.0.0e.orig/engines/ccgost/Makefile	2013-01-19 15:39:39 +0000
+++ openssl-1.0.0e/engines/ccgost/Makefile	2013-01-19 16:01:06 +0000
@@ -48,25 +48,25 @@
 		case "$(PLATFORM)" in \
 		"Cygwin" ) \
 			sfx=".so"; \
-			cp cyg$(LIBNAME).dll $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$${pfx}$(LIBNAME)$$sfx.new; \
+			cp cyg$(LIBNAME).dll $(INSTALL_PREFIX)$(ENGINESDIR)/$${pfx}$(LIBNAME)$$sfx.new; \
 			;; \
 		"MSys" ) \
 			sfx=".so"; \
-			cp msys-$(LIBNAME).dll $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$${pfx}$(LIBNAME)$$sfx.new; \
+			cp msys-$(LIBNAME).dll $(INSTALL_PREFIX)$(ENGINESDIR)/$${pfx}$(LIBNAME)$$sfx.new; \
 			;; \
 		*) \
 			case "$(CFLAGS)" in \
 			*DSO_BEOS*) sfx=".so";; \
 			*DSO_DLFCN*) sfx=`expr "$(SHLIB_EXT)" : '.*\(\.[a-z][a-z]*\)' \| ".so"`;; \
 			*DSO_DL*) sfx=".sl";; \
-			*DSO_WIN32*) sfx="eay32.dll"; pfx=;; \
+			*DSO_WIN32*) sfx=".dll";; \
 			*) sfx=".bad";; \
 			esac; \
-			cp $${pfx}$(LIBNAME)$$sfx $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$${pfx}$(LIBNAME)$$sfx.new; \
+			cp $${pfx}$(LIBNAME)$$sfx $(INSTALL_PREFIX)$(ENGINESDIR)/$${pfx}$(LIBNAME)$$sfx.new; \
 			;; \
 		esac; \
-		chmod 555 $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$${pfx}$(LIBNAME)$$sfx.new; \
-		mv -f $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$${pfx}$(LIBNAME)$$sfx.new $(INSTALL_PREFIX)$(INSTALLTOP)/$(LIBDIR)/engines/$${pfx}$(LIBNAME)$$sfx; \
+		chmod 555 $(INSTALL_PREFIX)$(ENGINESDIR)/$${pfx}$(LIBNAME)$$sfx.new; \
+		mv -f $(INSTALL_PREFIX)$(ENGINESDIR)/$${pfx}$(LIBNAME)$$sfx.new $(INSTALL_PREFIX)$(ENGINESDIR)/$${pfx}$(LIBNAME)$$sfx; \
 	fi
 
 links:
