--- openssl-1.0.0d/crypto/opensslconf.h.in	2005-12-16 13:37:23 +0300
+++ openssl-1.0.0d/crypto/opensslconf.h.in	2011-04-01 19:11:02 +0400
@@ -5,8 +5,8 @@
 
 #if !(defined(VMS) || defined(__VMS)) /* VMS uses logical names instead */
 #if defined(HEADER_CRYPTLIB_H) && !defined(OPENSSLDIR)
-#define ENGINESDIR "/usr/local/lib/engines"
-#define OPENSSLDIR "/usr/local/ssl"
+#define ENGINESDIR "/mingw/lib/openssl/engines-1.0.0"
+#define OPENSSLDIR "/mingw/var/ssl"
 #endif
 #endif
 
