--- util/strgutil.c.orig	Thu Dec 20 18:22:28 2012
+++ util/strgutil.c	Thu Apr 18 21:48:15 2013
@@ -138,7 +138,7 @@
       done = 1; /* Do it right now because we might get called recursivly
                    through gettext.  */
     
-      handle = dlopen ("iconv.dll", RTLD_LAZY);
+      handle = dlopen ("libiconv-2.dll", RTLD_LAZY);
       if (handle)
         {
           iconv_open  = dlsym (handle, "libiconv_open");
@@ -150,7 +150,7 @@
       if (!handle || !iconv_close)
         {
           log_info (_("error loading `%s': %s\n"),
-                     "iconv.dll",  dlerror ());
+                     "libiconv-2.dll",  dlerror ());
           log_info(_("please see http://www.gnupg.org/download/iconv.html "
                      "for more information\n"));
           iconv_open = NULL;
