--- openssl-0.9.8e/tools/c_rehash.in.orig	2007-08-27 15:10:34.000000000 -0300
+++ openssl-0.9.8e/tools/c_rehash.in	2007-08-27 15:12:41.000000000 -0300
@@ -59,7 +59,7 @@
 		}
 	}
 	closedir DIR;
-	FILE: foreach $fname (grep {/\.pem$/} @flist) {
+	FILE: foreach $fname (grep {/\.pem$|\.crt$/} @flist) {
 		# Check to see if certificates and/or CRLs present.
 		my ($cert, $crl) = check_file($fname);
 		if(!$cert && !$crl) {
